### DISCUSSION TITLE
How to use environments when used for secrets for non-deploying jobs but still want manual approval during deployment?

### DISCUSSION THREAD

**Discussion by Michaelvsk** [community/community#27600] [Actions] [answered]
Hi,

we are using GitHub Actions environments for both, to store env-specific secrets, and to integrate manual approval on deployment-jobs.
The problem is, we also need to access the secrets during build-jobs, which actually do not deploy anything and don't want have the approval-requirement there (e.g. for the planning step of a terraform deploylemt).

What we are doing at the moment is, to duplicate the environment mit their secrets and have one without approval-requirement and the other one with. E.g., `dev` and `dev-with-approval`.

Simple example:
```YAML
name: Deploy Infrastructure

on:
  workflow_dispatch:

jobs:
  terraform-plan:
    uses: <our_resuable_worflow_repo>/.github/workflows/terraform-plan.yaml@main
    with:
      github-environment: 'dev'
    secrets:
      TF_STATE_STORAGE_CONTAINER_NAME: ${{ secrets.TF_STATE_STORAGE_CONTAINER_NAME }}
      ARM_CLIENT_SECRET: ${{ secrets.ARM_CLIENT_SECRET }}
      # other secrets omitted for better readability

  terraform-apply:
    needs: terraform-plan
    uses: <our_resuable_worflow_repo>/.github/workflows/terraform-apply.yaml@main
    with:
      github-environment: 'dev-with-approval'
    secrets:
      TF_STATE_STORAGE_CONTAINER_NAME: ${{ secrets.TF_STATE_STORAGE_CONTAINER_NAME }}
      ARM_CLIENT_SECRET: ${{ secrets.ARM_CLIENT_SECRET }}
      # other secrets omitted for better readability
```

The problem is, that we
1. have to store the secrets redundantly in both environments,
2. face a messed up environments overview on the project homepage:
![image](https://user-images.githubusercontent.com/13169777/181729453-be1d037f-90c9-410b-b8f2-77cb37356864.png)

**Comment by Evovita** [comment, marked as answer]
**'Auto approve' workaround**

I have a similar problem whereby Approvals are required when manually run by users for the Production environment but at the same time we need _automated_ workflow requests e.g. via _GitHub API_ from _Jenkins_ to NOT be subject to approvals. I came up with the following workaround:

**Description**

1. Create an 'auto approve' job in your workflow that runs concurrently with the job that needs to be approved.
2. Add a condition to the 'auto approve' job to ensure it runs only when the automated workflow request user is the triggering actor (and if required, a condition that checks the desired target environment name e.g Prod).
3. In the 'auto approve' job, use the _GitHub API_ to [approve the pending deployment](https://docs.github.com/en/rest/actions/workflow-runs#review-pending-deployments-for-a-workflow-run). This works because by the time the concurrent 'auto approve' job executes it's code, the pending deployment state is available to be approved.
4. Add the user being auto approved to the list of Required Reviewers in the Environment settings.

**Pseudo code**

(Sorry, but my actual working code is in a private repo):
```yaml
...
jobs:

  first-job:
  
    name: First job
    runs-on: ubuntu-latest
    ...

  # IMPORTANT: Auto approve job must run concurrently with the job that requires approval so it detects the approval request.
  auto-approve-job:
    if: github.triggering_actor == 'username-running-automated-workflow-request'
    name: Auto approve job
    needs: first-job # Concurrency with Job that needs auto-approval
    runs-on: ubuntu-latest
      
    steps:
     
    - name: Auto approve
      shell: pwsh
      run: |        
        $url = "https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/pending_deployments";
        $body = @"
        { "environment_ids": [ 123456 ],
          "state": "approved",
          "comment": "Auto-approved for automation"
        }
        "@;
        Invoke-RESTMethod -Uri $url -Method POST -Headers @{
          Authorization = "Bearer ${{ secrets.MY_SECRET_PAT }}"
        } -Body $body -ContentType "application/json";
      

  job-that-requires-conditional-auto-approval:
    needs: first-job # Concurrency with auto approve job
    name: Job that requires conditional auto approval
    environment:
      name: Prod
  ...
```
If you want to use an Environment Name e.g. 'Prod' instead of hardcoding the `environment_ids` parameter, then you can find the environment id by [getting the list of Environments](https://docs.github.com/en/rest/deployments/environments#get-all-environments) using the _GitHub API_ . For example:

```yaml
...
    - name: Get Environment Id
      id: environment-id
      shell: pwsh
      run: |
        $environmentId = ""
        $url = "https://api.github.com/repos/${{ github.repository }}/environments"
        $response = Invoke-RestMethod -Uri $url -Method Get  -Headers @{
          Authorization = "Bearer ${{ secrets.MY_SECRET_PAT }}"
        }
        if ($response){
          # Find Environment Id...
          foreach ($environment in $response.environments) {
            if ($environment.name -eq "Prod"){
              $environmentId = $environment.id
              break
            }
          }
        }

        # Set output
        echo "::set-output name=value::$(echo $environmentId)"

    - name: Show me Environment Id
      run: |
        echo "Environment Id for Prod is ${{ steps.environment-id.outputs.value }}"
...
```

**Reply by Michaelvsk** [reply]
Great! Very nice idea to circumvent that issue. :) I will give it try, for sure.

Remark:
I will not mark it as an answer to see if there are other suggestions which are less "tricked". Maybe there is some easy to consume service for secret management, so environment are not needed to store them. If not, I'll mark it as answer after my vacation. ;)

**Reply by Evovita** [reply]
Thanks! Yes hopefully there is a better/simpler way of doing this. It seems such an essential part of CI/CD that I'm surprised GitHub have not already included e.g. an 'Auto-approve these users' list or similar.

**Comment by mcascone** [comment]
I have a similar case in which for a given environment, some repos can be set to auto-deploy, while other repos require manual approval. I think your pattern of using another job to auto-approve it, when appropriate, makes the most sense given the constraints.

i am thinking along the lines of:

```yaml
jobs:
  job-before-staging:
    # Whatever the job before the approval-maybe-required job is.
    # We just need this to anchor the next two jobs to run in parallel.

  # this env has deploy protection rules on it, must be manually approved
  deploy-staging:
    needs: job-before-staging
    env: STAGING
    if: runStaging == true
    # ... deploy steps here


  # no env declared here
  staging-auto-approval:
    needs: job-before-staging
    if: runStaging == true && requireStagingApproval == false

    steps:
      - name: auto-approve-deploy
        run: <the API to approve the job>
```

I'll try to remember to come back here with any findings. 🕵️

**Reply by mcascone** [reply]
So I've been fighting with this all day. Using the API, whether through `gh api` or `curl`, I can get the `env id`, the `job id`, and the `pending_approvals` of the current job. I can _not_ get the approval to work. 

This code works, to get pending deploys:
```sh
gh api -H "Accept: application/vnd.github+json"       \
  -H "X-GitHub-Api-Version: 2022-11-28"               \
  --verbose                                           \
  repos/$REPO/actions/runs/$RUNID/pending_deployments
```

☝️ returns a whole slew of json data, proving it can see the pending deploy.

This code to try to approve the pending deploy **fails**:

```sh
gh api -H "Accept: application/vnd.github+json"       \
  -H "X-GitHub-Api-Version: 2022-11-28"               \
  --verbose                                           \
  repos/$REPO/actions/runs/$RUNID/pending_deployments \
  --method POST                                       \
  -F "environment_ids[]=$ENVID"                       \
  -f "state=approved"                                 \
  -f "comment=Auto-Approved by GitHub Actions"      
```

Result:

```sh
{
  "comment": "Auto-Approved by GitHub Actions",
  "environment_ids": [
    11130481765
  ],
  "state": "approved"
}

...

{
  "message": "Validation Failed",
  "errors": "No pending deployment requests to approve or reject",
  "documentation_url": "https://docs.github.com/rest/actions/workflow-runs#review-pending-deployments-for-a-workflow-run",
  "status": "422"
}

* Request took 469.770458ms
gh: No pending deployment requests to approve or reject (Validation Failed)
```

I'm at the end of my rope with this, I've literally been fighting with it all day. I don't understand. I'm also using this perm block, with no effect:

```yaml
    env:
      GH_TOKEN: ${{ github.token }}

    permissions:
      checks: write
      deployments: write
      actions: write
```

**Reply by Evovita** [reply]
@mcascone Can you share your workflow file so I can compare it with mine?

**Reply by Evovita** [reply]
Also, I recently found this [Automate Environment Deployment Approval Github action](https://github.com/marketplace/actions/automate-environment-deployment-approval) that may provide equivalent functionality if you don't need/want to roll your own. Disclaimer: I haven't tried it, and I'm not the author.

**Reply by mcascone** [reply]
Thanks @Evovita! I also just stumbled on that action yesterday! [My findings from testing it](https://stackoverflow.com/questions/79044626/cant-approve-deployment-via-cli-rest-api-even-when-there-are-pending-deploys). (A lot of the content is the same as my previous post here; I cross-posted it. Scroll down to "updates after more testing".)

TL;DR: the 3rd-party action's author states, and I confirmed with testing, that the `github.token` ***can not*** be set to an appropriate permission config to approve deploys. It needs to be done with a PAT. I proved that with the action; it doesn't work with `github.token`, but it _does_ work with `secrets.A_PAT_WITH_REPO_PERMS` :tada:.

However! When I use the same exact PAT in my own workflow, OR in the CLI with either `curl` or `gh api`, **it doesn't work** 😡. So either I'm doing something wrong in my code, or there's an issue in the GH API backend that is preventing it from working, since I have proven the PAT is correct. (The former is of course always more likely.)

I have one more thing to test: a fine-grained token. Since GH is trying to move to those, maybe it's implemented differently behind the scenes.

Additionally: The 3rd-party action works, but it has a flaw IMHO: [it approves _every_ deploy that's waiting, _not just the ones in the run it was fired from_](https://github.com/activescott/automate-environment-deployment-approval/issues/60). Also, when configured according to the docs, [it seems to generate multiple approvals for a single deployment](https://github.com/activescott/automate-environment-deployment-approval/issues/59). I think this is due to the trigger being `on: deployment_status`, instead of just `on: deploy`. I've logged two issues in the action's repo for these, but the action was last updated about 1.5 years ago, so I'm not sure how actively it's supported any more.

# Update after writing all of this
I realized that I'm _not_ using the `triggering_actor` condition you describe. Maybe that's the issue?

## Workflow
This is my workflow with my own API code:
<details>
<summary>click to expand</summary>

```yaml
name: conditions tester

on:
  push:
  pull_request:
  workflow_dispatch:

jobs:
  init:
    runs-on: ubuntu-latest

    outputs: 
      runDev: ${{ steps.set_output.outputs.runDev }}
      runProd: ${{ steps.set_output.outputs.runProd }}
      devApprovalNeeded: ${{ steps.set_output.outputs.devApprovalNeeded }}

    steps:
      - name: Run tests
        run: echo "Running init"

      - name: set output
        id: set_output
        run: |
          echo "runDev=true" >> $GITHUB_OUTPUT
          echo "runProd=true" >> $GITHUB_OUTPUT
          echo "devApprovalNeeded=false" >> $GITHUB_OUTPUT
  
  deployDev:
    runs-on: ubuntu-latest
    name: Deploy to DEV
    environment: DEV
    needs: init

    if: needs.init.outputs.runDev == 'true'
    
    steps:
      - name: deploy dev
        run: echo "Deploying to DEV" >> $GITHUB_STEP_SUMMARY
  
  deployDevAutoApprove:
    runs-on: ubuntu-latest
    name: Deploy to DEV Auto Approve
    needs: init

    env:
      GH_TOKEN: ${{ secrets.GH_DEPLOY_PAT }}
      # GH_TOKEN: ${{ github.token }} doesn't work

    # permissions: write-all # not needed with PAT

    if: needs.init.outputs.runDev == 'true' && needs.init.outputs.devApprovalNeeded != 'true'

    steps:
      - run: echo "Deploy to DEV Auto Approved" >> $GITHUB_STEP_SUMMARY
      
      - name: Get Environment Id
        id: environment-id
        run: |
          envs=$(gh api repos/${{ github.repository }}/environments)

          environmentId=$(echo $envs | jq -r '.environments[] | select(.name=="DEV") | .id')
          echo "ENVID=$environmentId" >> $GITHUB_ENV
      
      - name: Get Pending Deployments
        id: pending-deployments
        run: |
          gh api /repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/pending_deployments

		  # just forcing auto-approve here, I improved the check later
          echo "pending=true" >> $GITHUB_OUTPUT
      
      - name: Approve Pending Deployment
        if: steps.pending-deployments.outputs.pending != '[]'
        run: |
          gh api --method POST                            \
            -F "environment_ids[]=$ENVID"                 \
            -f "state=approved"                           \
            -f "comment=Auto-Approved by GitHub Actions"  \
            --verbose \
            repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/pending_deployments

  deployProd:
    runs-on: ubuntu-latest
    name: Deploy to PROD
    environment: PRD
    needs: [init, deployDev]

    if: ${{ needs.init.outputs.runProd == 'true' }}

    steps:
      - name: deploy prod
        run: echo "Deploying to PRD" >> $GITHUB_STEP_SUMMARY
```
</details>

**Comment by Evovita** [comment]
@mcascone Thanks for the info. I [copied your workflow](https://github.com/Evovita/test/blob/master/.github/workflows/Test%20Auto%20approve.yml) made some minor formatting changes and removed `echo "pending=true" >> $GITHUB_OUTPUT` - [this run](https://github.com/Evovita/test/actions/runs/11166338894) confirms the auto-deploy functionality works for me (the `PRD` environment approval was done manually) - I watched it switch state from waiting approval on the `DEV` environment to auto-approved:

![image](https://github.com/user-attachments/assets/b6aa8bed-106a-4f97-abed-a0cb73419230)

![image](https://github.com/user-attachments/assets/2f79689c-4b48-442f-ad4a-9b93f3c4124c)

### GH_DEPLOY_PAT fine-grained token settings used

![image](https://github.com/user-attachments/assets/9765284d-2915-4ed3-b476-2c7cd405c378)

### DEV environment config

![image](https://github.com/user-attachments/assets/50904566-f1aa-4d4b-ba24-cc05ba05854a)

**Reply by mcascone** [reply]
Thank you for this! I tried it again, and even only using my PAT, it works! I don't think I changed anything. One thing I might have shot myself in the foot with, was that the ENV ID I was using on my local CLI to test was wrong. It wouldn't be the first time I've lost so much time to a dumb mistake like that. But I was having the same problem in the real Actions context too.

I don't know. It works today. I'd love to know there was a problem on GitHub's backend yesterday that caused my issues, but it's much more likely that I just caught myself up in troubleshooting with invalid data.

FWIW it still does _not_ work with the default `github.token`. It only works with a PAT or fine-grained token. My PAT only has `repo` perms.

Thanks again for your effort!!
